File: ShrutiVarshney_BanasthaliVidyapith_BitManipulationAndGreedyAssignment/BitManipulationAndGreedyAssignmentQuestions/src/bitmanipulationandgreedyassignmentquestions/Problem6.java
package bitmanipulationandgreedyassignmentquestions;
import java.util.*;
public class Problem6 
{
    public static int findSingleNumber(int[] nums) 
    {
        int result = 0;
        for (int i = 0; i < 32; i++) 
        {
            int count = 0;
            for (int num : nums) 
            {
                // Count the number of elements with a set bit at the i-th position
                if ((num & (1 << i)) != 0) {
                    count++;
                }
            }
            // If the count is not a multiple of 3, the single number has a set bit at the i-th position
            if (count % 3 != 0) 
            {
                result |= (1 << i);
            }
        }
        return result;
    }

    public static void main(String[] args) 
    {
        int[] nums1 = {1, 2, 4, 3, 3, 2, 2, 3, 1, 1};
        int singleNumber1 = findSingleNumber(nums1);
        System.out.println(singleNumber1); // Output: 4

        int[] nums2 = {0, 0, 0, 1};
        int singleNumber2 = findSingleNumber(nums2);
        System.out.println(singleNumber2); // Output: 1
    }
}

File: student_Banasthali_Nivedita/bitManipulation_and_Greedy/problem6.cpp
/*
Single Number
*/

#include<bits/stdc++.h>
using namespace std;

int singleNumber(const vector<int>& A) {
    int result = 0;

    for (int i = 0; i < 32; i++) {
        int count = 0;

        // Count the number of occurrences of the bit at position i
        for (int num : A) {
            if ((num >> i) & 1) {
                count++;
            }
        }

        // If the count is not divisible by 3, set the bit in the result
        if (count % 3 != 0) {
            result |= (1 << i);
        }
    }

    return result;
}

int main() {
    vector<int> A;

    // Example 1
    A = {1, 2, 4, 3, 3, 2, 2, 3, 1, 1};
    int result1 = singleNumber(A);
    cout << "Example 1: " << result1 << endl;
    
    return 0;
}

